s3/src/main/scala/monix/connect/s3/MultipartUploadSubscriber.scala

```diff
@@ -0,0 +1,185 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.connect.s3
+
+import monix.eval.Task
+import monix.execution.cancelables.AssignableCancelable
+import monix.execution.{Ack, Callback, Scheduler}
+import monix.reactive.Consumer
+import monix.reactive.observers.Subscriber
+import software.amazon.awssdk.core.async.AsyncRequestBody
+import software.amazon.awssdk.services.s3.S3AsyncClient
+import software.amazon.awssdk.services.s3.model._
+import scala.util.control.NonFatal
+
+/**
+  * An Amazon S3 multipart uploads is an write operation made progresively on individual chunks of an object (parts),
+  * that finishes with a message to S3 that multipart upload is completed, in which the concatenation
+  * of each of the individual chunks will end up stored into the same object.
+  * Implement the functionality to perform an S3 multipart update from received chunks of data.
+  */
+private[s3] class MultipartUploadSubscriber(
+  bucket: String,
+  key: String,
+  minChunkSize: Int = MultipartUploadSubscriber.awsMinChunkSize,
+  acl: Option[String],
+  contentType: Option[String] = None,
+  grantFullControl: Option[String],
+  grantRead: Option[String],
+  grantReadACP: Option[String],
+  grantWriteACP: Option[String],
+  serverSideEncryption: Option[String] = None,
+  sseCustomerAlgorithm: Option[String] = None,
+  sseCustomerKey: Option[String] = None,
+  sseCustomerKeyMD5: Option[String] = None,
+  ssekmsEncryptionContext: Option[String],
+  ssekmsKeyId: Option[String],
+  requestPayer: Option[String])(
+  implicit
+  s3Client: S3AsyncClient)
+  extends Consumer.Sync[Array[Byte], CompleteMultipartUploadResponse] {
+
+  private val createRequest: CreateMultipartUploadRequest =
+    S3RequestBuilder.createMultipartUploadRequest(
+      bucket = bucket,
+      key = key,
+      contentType = contentType,
+      acl = acl,
+      grantFullControl = grantFullControl,
+      grantRead = grantRead,
+      grantReadACP = grantReadACP,
+      grantWriteACP = grantWriteACP,
+      requestPayer = requestPayer,
+      serverSideEncryption = serverSideEncryption,
+      sseCustomerAlgorithm = sseCustomerAlgorithm,
+      sseCustomerKey = sseCustomerKey,
+      sseCustomerKeyMD5 = sseCustomerKeyMD5,
+      ssekmsEncryptionContext = ssekmsEncryptionContext,
+      ssekmsKeyId = ssekmsKeyId
+    )
+
+  private val uploadId: Task[String] =
+    Task
+      .from(s3Client.createMultipartUpload(createRequest))
+      .map(_.uploadId())
+      .memoize
+
+  private var buffer: Array[Byte] = Array.emptyByteArray
+
+  private var completedParts: List[CompletedPart] = List.empty[CompletedPart]
+
+  private var partN = 1
```

Comment:
What happens with these vars if someone reuses `Consumer` in two places? No idea if it is a valid concern but it might make sense to move them to `createSubscriber`

Reply:
that's most likely to happen, will sort that